### PR TITLE
ObjectLoader: Fix order of loading events

### LINKED ISSUE #12601: Wrong order of loading events
I am loading a scene with textures via ObjectLoader.

Live test: http://shodiev.ru/other/three_events/
Zip with testcase: https://yadi.sk/d/sGqOqpbm3PVSM9

As you can see in console, events are firing in unexpected order: function that i pass to LoadingManager is called before ObjectLoader is complete.

It looks like LoadingManager completes when only _scene.json_ is loaded although there's still images to load.

Is it supposed to be like that?

##### Three.js version

- [ ] Dev
- [x] r88
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by Itee**
This PR fix #12601 

All example run correctly, and image are correctly loaded before the call to onLoad. 
It simplify the code, too, avoiding chaining onLoad parameters too deeper.

### CODE REVIEW

File: src/loaders/ObjectLoader.js [hunk 1]
```diff
@@ -142,9 +138,18 @@ Object.assign( ObjectLoader.prototype, {
...
 		}
 
-		if ( json.images === undefined || json.images.length === 0 ) {
+		// fix #12601
+		// Wrap LoadingManager onLoad methods with ObjectLoader onLoad callback,
+		// allow proper callback call order
+		if ( onLoad !== undefined ) {
+
+			var loadingManagerOnLoad = scope.manager.onLoad;
+			scope.manager.onLoad = function warpedOnLoad() {
```
Comment (mrdoob) [inline src/loaders/ObjectLoader.js]: Hmm, everything looks great except this part...
The only "pretty" solution I can think of is making `LoadingManager` extend `EventDispatcher`.

Comment (Itee) [reply src/loaders/ObjectLoader.js]: Yes this is the point of interest of this PR...
Extend LoadingManager with EventDispatcher will allow to listen event like:

```javascript
if ( onLoad !== undefined ) {

	scope.manager.addEventListener( 'onLoad', function () {

		onLoad( object );

	} );

}
```

with inside LoadingManager things like:

```javascript
this.itemEnd = function ( url ) {

	itemsLoaded ++;

	if ( scope.onProgress !== undefined ) {

		scope.dispatchEvent( { type: 'onProgress' } );		// Note here
		scope.onProgress( url, itemsLoaded, itemsTotal );

	}

	if ( itemsLoaded === itemsTotal ) {

		isLoading = false;

		if ( scope.onLoad !== undefined ) {

			scope.dispatchEvent( { type: 'onLoad' } );	// Note here
			scope.onLoad();

		}

	}

};
```

Or maybe events should be called: beforeOnProgress, beforeOnLoad, beforeOnError which are more just about the underlying meaning...

Comment (Itee) [reply src/loaders/ObjectLoader.js]: Should i make an other PR about the LoadingManager extends ?

**Comment by Itee** [comment]
@mrdoob I extend LoadingManager with EventDispatcher like you said. I fire the basic events with the minimum of data to be usable. And fix the "wrap" listening the end event.

This is prettier than the previous version !

**Pull request closed without merging**